pyomo/contrib/solver/solvers/gurobi/gurobi_direct.py

```diff
@@ -0,0 +1,134 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2025
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+import operator
+
+from pyomo.common.collections import ComponentMap, ComponentSet
+from pyomo.common.shutdown import python_is_shutting_down
+from pyomo.core.staleflag import StaleFlagManager
+from pyomo.repn.plugins.standard_form import LinearStandardFormCompiler
+
+from pyomo.contrib.solver.common.util import (
+    NoDualsError,
+    NoReducedCostsError,
+    NoSolutionError,
+    IncompatibleModelError,
+)
+from pyomo.contrib.solver.common.solution_loader import SolutionLoaderBase
+from .gurobi_direct_base import (
+    GurobiDirectBase,
+    gurobipy,
+    GurobiDirectSolutionLoaderBase,
+)
+import logging
+
+
+logger = logging.getLogger(__name__)
+
+
+class GurobiDirectSolutionLoader(GurobiDirectSolutionLoaderBase):
+    def __del__(self):
+        super().__del__()
+        if python_is_shutting_down():
+            return
+        # Free the associated model
+        if self._solver_model is not None:
+            self._var_map = None
+            self._con_map = None
+            # explicitly release the model
+            self._solver_model.dispose()
+            self._solver_model = None
+
+
+class GurobiDirect(GurobiDirectBase):
+    _minimum_version = (9, 0, 0)
+
+    def __init__(self, **kwds):
+        super().__init__(**kwds)
+        self._gurobi_vars = None
+        self._pyomo_vars = None
```

Comment:
Why are these being stored on the solver interface?  Direct is supposed to be non-persistent, and adding this seems to violate that principle.

Reply:
Is this going to work for callbacks too? They have the same issue, that they need this state (and actually more) to work... So I think there may be a bigger design question lurking here: For example, they need the gurobipy model to be somewhere they can get to: In persistent, that is currently also sitting on the solver interface.

Reply:
(I was originally thinking that this kind of state could be added in a contextmanager kind of style, and removed before `solve` is over, regardless of what happens. But that certainly might not be the right answer.)